### PR TITLE
JDK-8296907: VMError: add optional callstacks, siginfo for secondary errors

### PR DESCRIPTION
**Pull request by tstuefe**
This was motivated by discussions we had in https://github.com/openjdk/jdk/pull/11017.

To aid in analyzing secondary errors during error reporting, it would be useful to see their callstacks for secondary errors. But printing callstacks during error reporting is unsafe - if we get a second crash or assert, it will cause infinite recursion and interrupt error reporting. Also, the hs-err file would be quite verbose. Therefore this feature is optional and limited to debug builds. 

---

Patch

- adds optional callstack/siginfo printing via debug-only switch `-XX:+ErrorLogSecondaryErrorDetails`.
- fixes a bug in secondary error handling where we would use the global scratch buffer recursively (via stringStream); that could lead to confusing output since it is used by the error log stream already. We can print directly to that one instead.
- Removed a stray newline from print_native_stack to clean output.
- added regression testing for this feature. I removed my name and the bug id from the test since we don't do this anymore (?).
- added clarifying comments to the test and code
- added SAP copyright to the regression test (we introduced it years ago for JDK-8065895)

Output looks like this:

```
$ java ... -XX:+ErrorLogSecondaryErrorDetails
```

will produce, for secondary errors, siginfo and call stack.

```
[error occurred during error reporting (test secondary crash 1), id 0xb, SIGSEGV (0xb) at pc=0x00007fddfe8a0a61]
[siginfo: si_signo: 11 (SIGSEGV), si_code: 128 (SI_KERNEL), si_addr: 0x0000000000000000]
[stack: Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x1ceea61]  VMError::controlled_crash(int)+0x241  (vmError.cpp:1946)
V  [libjvm.so+0x1cf413f]  VMError::report(outputStream*, bool)+0x46bf  (vmError.cpp:564)
V  [libjvm.so+0x1cf516b]  VMError::report_and_die(int, char const*, char const*, __va_list_tag*, Thread*, unsigned char*, void*, void*, char const*, int, unsigned long)+0x19b  (vmError.cpp:1709)
V  [libjvm.so+0x1cf5e8f]  VMError::report_and_die(Thread*, unsigned int, unsigned char*, void*, void*, char const*, ...)+0x8f  (vmError.cpp:1467)
V  [libjvm.so+0x1cf5ec2]  VMError::report_and_die(Thread*, unsigned int, unsigned char*, void*, void*)+0x22  (vmError.cpp:1473)
V  [libjvm.so+0x1a549e7]  JVM_handle_linux_signal+0x1f7  (signals_posix.cpp:656)
C  [libc.so.6+0x43090]
V  [libjvm.so+0x11d6965]  JNI_CreateJavaVM+0x5b5  (jni.cpp:3662)
C  [libjli.so+0x4013]  JavaMain+0x93  (java.c:1457)
C  [libjli.so+0x800d]  ThreadJavaMain+0xd  (java_md.c:650)
]
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296907](https://bugs.openjdk.org/browse/JDK-8296907): VMError: add optional callstacks, siginfo for secondary errors

### Reviewers
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - Committer)
 * [Ralf Schmelter](https://openjdk.org/census#rschmelter) (@schmelter-sap - **Reviewer**) ⚠️ Review applies to [ae92af84](https://git.openjdk.org/jdk/pull/11118/files/ae92af8494eb4df23fc01876df82632ed280d09f)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11118/head:pull/11118` \
`$ git checkout pull/11118`

Update a local copy of the PR: \
`$ git checkout pull/11118` \
`$ git pull https://git.openjdk.org/jdk pull/11118/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11118`

View PR using the GUI difftool: \
`$ git pr show -t 11118`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11118.diff">https://git.openjdk.org/jdk/pull/11118.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/runtime/ErrorHandling/SecondaryErrorTest.java
Comment (turbanoff) [inline test/hotspot/jtreg/runtime/ErrorHandling/SecondaryErrorTest.java:159]:
```suggestion
      throw new RuntimeException("hs-err file incomplete (first missing pattern: " + pattern[currentPattern] + ")");
```

File: src/hotspot/share/utilities/vmError.cpp
Comment (dholmes-ora) [inline src/hotspot/share/utilities/vmError.cpp:1631]: Suggestion: "Print brief information, but ..."

File: test/hotspot/jtreg/runtime/ErrorHandling/SecondaryErrorTest.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ErrorHandling/SecondaryErrorTest.java:72]: Nit: include the unknown argument in the message

**Review by dholmes-ora — COMMENTED** [review]
Seems reasonable in terms of the code, but I am curious how you expect this to be used in practice? If you get a secondary crash with this turned off then I would expect to use gdb on the core file to get the secondary stack, rather than using the new flag and hoping to reproduce the problem. To be useful it seems you would have to remember to run with this always enabled when developing/debugging. ??

**Comment by tstuefe** [comment]
> Seems reasonable in terms of the code, but I am curious how you expect this to be used in practice? If you get a secondary crash with this turned off then I would expect to use gdb on the core file to get the secondary stack, rather than using the new flag and hoping to reproduce the problem. To be useful it seems you would have to remember to run with this always enabled when developing/debugging. ??

Yes, sure. It is easier than getting the code and starting gdb, especially if you have no access to the core.

If the secondary error is reproducible, and they often are, and the developer's goal is to fix them, he sets the option. Even if they are not reproducible - if people get annoyed enough, we can just enable it for tests where they regularly happen. The option does not cost anything in terms of normal runtime. We also could always enable it in runtime/ErrorHandling tests. At SAP, we even have test specifically for sniffing out and fixing secondary crashes, but since they are work-intensive in practice we never contributed them.

Not sure if you followed https://github.com/openjdk/jdk/pull/11017. Here and in other RFEs we argued about granularity of error-reporting STEPs with their seconary-signal-catch capability. There were voices for making them more fine-grained. But that would cause other problems, and therefore it would be good to harden out the error reporting code instead. This RFE tries to make this work a little less onerous.

File: src/hotspot/share/utilities/vmError.cpp
Comment (xmas92) [inline src/hotspot/share/utilities/vmError.cpp:1635]: Is there any problems with making this static? Given that we care about the stack depth when we have repeated crashes.

File: src/hotspot/share/utilities/vmError.cpp
Comment (xmas92) [inline src/hotspot/share/utilities/vmError.cpp:1653]: > But printing callstacks during error reporting is unsafe - if we get a second crash or assert, it will cause infinite recursion and interrupt error reporting.

Should there be logic here that at least guarantee progress?
```c++
if (ErrorLogSecondaryErrorDetails && !_some_bool) {
  _some_bool = true;
  [...]
}
_some_bool = false;
```

File: src/hotspot/share/utilities/vmError.cpp
Comment (xmas92) [inline src/hotspot/share/utilities/vmError.cpp:1641]: Not really relevant for this PR. But do not like the inconsistency that the id is printed as hex here, decimal in os::print_siginfo for posix, and hex in os::print_siginfo for windows.

**Review by xmas92 — COMMENTED** [review]
Always nice with more, simple and unobtrusive development tooling. 
Just have a couple of comments/questions?

File: src/hotspot/share/utilities/vmError.cpp
Comment (tstuefe) [reply src/hotspot/share/utilities/vmError.cpp:1635]: I think not. At this point we are single threaded. I'll convert to static.

File: src/hotspot/share/utilities/vmError.cpp
Comment (tstuefe) [reply src/hotspot/share/utilities/vmError.cpp:1653]: It would work horizontally too, though, since it would disable call stack printing for the next secondary error.

I still think yours is a good idea. We can extend it to using a counter and only allow a small number (2-5) of secondary errors to get printed out with their call stacks. Because if we have more than, say, 5 recursive errors, chances are they all trip over the same thing anyway and more call stacks won't tell you anything new.

And then, if we limit the number of these things, maybe we can afford to leave this feature on by default in debug builds. That addresses the concern  @dholmes-ora voiced, about having to switch on the feature first.

File: src/hotspot/share/utilities/vmError.cpp
Comment (tstuefe) [reply src/hotspot/share/utilities/vmError.cpp:1641]: I think the whole concept of VMError::id is not that good. It holds either SEH codes (windows), signal numbers or some enum we ourselves define. So it combines several orthogonal things, and the encoding is not clear. I have several half-done patches laying around cleaning it up, but never got around to finish them. I think some hierarchical structure would be better, e.g. errortype="crash|assert|oom", with auxilliary information depending on errortype attached. The encoding can be as dense as a 32-bit value, if one wanted. But that would also sort the printing problem, since printing depends on what the id is in the first place (e.g. signals are usually printed with d, but SEH codes as 32bit hex).

File: src/hotspot/share/utilities/vmError.cpp
Comment (xmas92) [reply src/hotspot/share/utilities/vmError.cpp:1653]: >  since it would disable call stack printing for the next secondary error.

 Would it? The example above would only disable reentering the if scope if we have a crash in `[...]`. If it is fine the flag is reset, and if we crash in `[...]` it will not enter the if scope and reset the flag. When it crash in a new secondary error it will once again enter the if scope and try to print the call stack. 

But I agree that there is always this taking it one step deeper of where you make things recoverable.

Comment (tstuefe) [reply src/hotspot/share/utilities/vmError.cpp:1653]: Right, you did reset it, I see now. Sorry, did not look close enough :)

**Comment by tstuefe** [comment]
Followed Axels idea of limiting recursion errors when printing secondary error crashes. I briefly played with more changes, and optionally removing the new switch, but in the end kept the change simple.

I also disabled source code printing from the secondary crash call stacks since that may be quite fragile.

**Comment by tstuefe** [comment]
While playing with this, I thought this would be a lot better with 
- limited stack size, either up to the before-last level of VMError::report invocation, or simply just n frames. Otherwise with subsequent crashes stacks get ridiculously long.
- Maybe enabling the switch by default in debug builds? And in addition limit number of secondary crash callstacks to the first five.

Subsequent ideas wrt recursive error limit: I think we could probably expand it.

File: src/hotspot/share/utilities/vmError.cpp
Comment (xmas92) [inline src/hotspot/share/utilities/vmError.cpp:355]: Only thing is that we may now print more than StackPrintLimit. Maybe clamping max_frames is more appropriate. 

But I do not see a reason why someone would lower StackPrintLimit and use ErrorLogSecondaryErrorDetails.

**Review by xmas92 — APPROVED** [review]
lgtm.

**Comment by tstuefe** [comment]
@dholmes-ora do you have any objections to this? If you lack time, that is fine, then I try to get another reviewer.

File: src/hotspot/share/utilities/vmError.cpp
Comment (schmelter-sap) [inline src/hotspot/share/utilities/vmError.cpp:1527]: Same as above. Would not work if we have parallel crashes, but I don't think it warrants making the code more complicated to catch this.

**Review by schmelter-sap — APPROVED** [review]
Loks good to me. I woulld really advocate to make it usable in product builds.

File: src/hotspot/share/utilities/vmError.cpp
Comment (tstuefe) [reply src/hotspot/share/utilities/vmError.cpp:1507]: I'm not sure how this would happen: we only end up here if we are the original thread that triggered the first fault (_first_error_tid == me). Otherwise, we would end up in `os::infinite_sleep()` at lines 1455ff. If a different thread crashes inside signal handling, between signal handler start and infinite_sleep, it will just keep crashing endlessly and run out of stack.

**Comment by tstuefe** [comment]
Hi @schmelter-sap, thanks a lot for the review! I made the flag a product flag as you suggested. Is this version still okay?

@xmas92 @dholmes-ora barring any objections, I plan to push tomorrow if all tests tonight are green.

Cheers, Thomas

**Comment by tstuefe** [comment]
Our nightly tests are green. GHA errors unrelated. I re-ran tests on 32-bit x86, all good. If I don't hear any objections, I will integrate later today.

**Comment by dholmes-ora** [comment]
No objections from me, we will see how this works out in practice.

**Review by xmas92 — APPROVED** [review]

**Comment by tstuefe** [comment]
Thanks @xmas92 and @schmelter-sap for reviewing!

/integrate

**Pull request closed without merging**
